### PR TITLE
Add  support for int to string enumeration

### PR DESCRIPTION
**Pull request by oliver-walker-hms**

### CODE REVIEW

File: src/com/hms_networks/americas/sc/taginfo/TagType.java [hunk 1]
```diff
@@ -23,6 +23,9 @@
   /** Ewon integer value for string type. */
   private static final int STRING_INT = 6;
 
+  /** Ewon integer value for mapped an integer mapped to a string. */
```
Comment (alexjhawk) [inline src/com/hms_networks/americas/sc/taginfo/TagType.java:26]: I think adding the new tag type in TagType.java should be a separate commit

File: src/com/hms_networks/americas/sc/taginfo/TagInfoManager.java [hunk 2]
```diff
@@ -282,24 +302,168 @@ private static synchronized void processTagListEBDLine(String line) {
...
+    JSONArray tagEnumJsonArray;
+    if (enableIntToStringEnums) {
+      allTagEnumJson =
+          new JSONObject(FileAccessManager.readFileToString(INT_TO_STRING_ENUMERATION_FILE_NAME));
+      tags = (JSONObject) allTagEnumJson.get("tags");
+      tagEnumJsonArray = (JSONArray) tags.get(tagName);
+      final int length = tagEnumJsonArray.length();
+      int biggest = 0;
+
+      /* enumerations are held in a JSON array of one key/value pair for each index in the array.
```
Comment (alexjhawk) [inline src/com/hms_networks/americas/sc/taginfo/TagInfoManager.java:442]:
```suggestion
      /* Enumerations are held in a JSON array of one key/value pair for each index in the array.
```

**Review by alexjhawk — CHANGES_REQUESTED** [review]

**Review by alexjhawk — APPROVED** [review]
Content/code seems good, although I would suggest cleaning up so that we only really use the terminology "integer mapped string". Some places refer to it like that, and others say enumeration.

**Review by it-hms — DISMISSED** [review]

**Comment by oliver-walker-hms** [comment]
I've changed a few comments to say "int to string enumeration" instead of "enumeration", as well as added "intToStringEnumeration" to a few variable names that had "enumeration"

File: src/com/hms_networks/americas/sc/taginfo/package.html [hunk 3]
```diff
@@ -3,8 +3,13 @@
 Ewon Flexy tag information library created by the HMS Networks MU Americas Solution Center.
 This library is dependent on the HMS Networks MU Americas Solution Center library for strings,
 found at https://www.github.com/hms-networks/sc-string-lib.
+This library is dependent on the HMS Networks MU Americas Solution Center library for file
+utilities, found at https://github.com/hms-networks/sc-flexy-file-util-lib.
+This library is dependent on the HMS Networks MU Americas Solution Center library for JSON
+management, found at https://github.com/hms-networks/sc-flexy-json-lib.
 
-@version 1.1
+
```
Comment (alexjhawk) [inline src/com/hms_networks/americas/sc/taginfo/package.html:11]: extra line was added -- remove

**Review by alexjhawk — CHANGES_REQUESTED** [review]

File: src/com/hms_networks/americas/sc/taginfo/TagInfoManager.java [hunk 4]
```diff
@@ -282,24 +302,168 @@ private static synchronized void processTagListEBDLine(String line) {
...
+    JSONObject allTagEnumJson;
+    JSONObject tags;
+    JSONArray tagEnumJsonArray;
+    if (enableIntToStringEnums) {
+      allTagEnumJson =
+          new JSONObject(FileAccessManager.readFileToString(INT_TO_STRING_ENUMERATION_FILE_NAME));
+      tags = (JSONObject) allTagEnumJson.get("tags");
+      tagEnumJsonArray = (JSONArray) tags.get(tagName);
+      final int length = tagEnumJsonArray.length();
+      int biggest = 0;
```
Comment (TomKimsey) [inline src/com/hms_networks/americas/sc/taginfo/TagInfoManager.java:440]: biggest what?

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,6 +1,9 @@
 # Ewon Flexy Tag Info Library Changelog
 
 ## v1.2
+### Major Changes
+- Feature: Add support for int to string tag enumerations
```
Comment (TomKimsey) [inline CHANGELOG.md:5]:
```suggestion
- Feature: Added support for int to string tag enumerations
```

File: src/com/hms_networks/americas/sc/taginfo/TagInfo.java [hunk 6]
```diff
@@ -30,6 +30,9 @@
   /** List of tag groups */
   private final ArrayList tagGroups;
 
+  /** Integer tags can be used to represent string values in a string mapping */
+  private final String[] enumeratedStringValueMapping;
```
Comment (TomKimsey) [inline src/com/hms_networks/americas/sc/taginfo/TagInfo.java:34]: Since now TagInfo contains type specific information would it make sense to break it out to

TagInfoInt
TagInfoBool
etc
then just have them inherit from TagInfo?

Seems strange to add a field here for just one special type of tag.

File: src/com/hms_networks/americas/sc/taginfo/TagInfo.java [hunk 7]
```diff
@@ -30,6 +30,9 @@
   /** List of tag groups */
   private final ArrayList tagGroups;
 
+  /** Integer tags can be used to represent string values in a string mapping */
```
Comment (TomKimsey) [inline src/com/hms_networks/americas/sc/taginfo/TagInfo.java:33]: did you mean to say "Integer values can ..."

File: src/com/hms_networks/americas/sc/taginfo/TagInfoManager.java [hunk 8]
```diff
@@ -282,24 +302,168 @@ private static synchronized void processTagListEBDLine(String line) {
...
+      for (int i = 0; i < length; i++) {
+        String value = tagListJsonArray.getString(i);
+        tagList.add(value);
+      }
+      intToStringEnumerationTagList = tagList;
+    }
+  }
+
+  /**
+   * Gets the int to string tag enumerations for a single tag.
```
Comment (TomKimsey) [inline src/com/hms_networks/americas/sc/taginfo/TagInfoManager.java:421]: This isnt clear what is being returned

File: src/com/hms_networks/americas/sc/taginfo/TagInfoManager.java [hunk 9]
```diff
@@ -282,24 +302,168 @@ private static synchronized void processTagListEBDLine(String line) {
...
+    String[] tagList = null;
+    JSONObject allTagEnumJson;
+    JSONObject tags;
+    JSONArray tagEnumJsonArray;
+    if (enableIntToStringEnums) {
+      allTagEnumJson =
+          new JSONObject(FileAccessManager.readFileToString(INT_TO_STRING_ENUMERATION_FILE_NAME));
+      tags = (JSONObject) allTagEnumJson.get("tags");
+      tagEnumJsonArray = (JSONArray) tags.get(tagName);
+      final int length = tagEnumJsonArray.length();
```
Comment (TomKimsey) [inline src/com/hms_networks/americas/sc/taginfo/TagInfoManager.java:439]: length of what?

File: src/com/hms_networks/americas/sc/taginfo/TagInfoManager.java [hunk 10]
```diff
@@ -282,24 +302,168 @@ private static synchronized void processTagListEBDLine(String line) {
...
+      }
+      tagList = new String[biggest + 1];
+      for (int i = 0; i < length; i++) {
+        JSONObject jsonObject = (JSONObject) tagEnumJsonArray.get(i);
+        String mappingKey = (String) jsonObject.names().get(tagEnumStringKeyIndex);
+        String value = (String) jsonObject.get(mappingKey);
+        tagList[Integer.parseInt(mappingKey)] = value;
+      }
+    }
+    return tagList;
```
Comment (TomKimsey) [inline src/com/hms_networks/americas/sc/taginfo/TagInfoManager.java:464]: why is this named tagList?

File: src/com/hms_networks/americas/sc/taginfo/TagInfoManager.java [hunk 11]
```diff
@@ -282,24 +302,168 @@ private static synchronized void processTagListEBDLine(String line) {
...
+   */
+  private static void getIntStringEnumTags() throws IOException, JSONException {
+    if (enableIntToStringEnums) {
+      ArrayList tagList = new ArrayList();
+      JSONObject allTagEnumJson;
+      JSONArray tagListJsonArray;
+      allTagEnumJson =
+          new JSONObject(FileAccessManager.readFileToString(INT_TO_STRING_ENUMERATION_FILE_NAME));
+      tagListJsonArray = (JSONArray) allTagEnumJson.get("enumeratedTagList");
+      final int length = tagListJsonArray.length();
```
Comment (TomKimsey) [inline src/com/hms_networks/americas/sc/taginfo/TagInfoManager.java:411]: length of what?

File: src/com/hms_networks/americas/sc/taginfo/TagInfoManager.java [hunk 12]
```diff
@@ -282,24 +302,168 @@ private static synchronized void processTagListEBDLine(String line) {
...
+   */
+  private static String[] getIntStringEnumTagMappings(String tagName)
+      throws IOException, JSONException {
+    String[] tagList = null;
+    JSONObject allTagEnumJson;
+    JSONObject tags;
+    JSONArray tagEnumJsonArray;
+    if (enableIntToStringEnums) {
+      allTagEnumJson =
+          new JSONObject(FileAccessManager.readFileToString(INT_TO_STRING_ENUMERATION_FILE_NAME));
```
Comment (TomKimsey) [inline src/com/hms_networks/americas/sc/taginfo/TagInfoManager.java:436]: Is there documentation on what this file format is?

**Review by TomKimsey — CHANGES_REQUESTED** [review]

**Comment by TomKimsey** [comment]
I dont think i have all the information i need to properly review this and the commit messages are not helpful in helping me understand these changes or why they are being made.

File: src/com/hms_networks/americas/sc/taginfo/TagInfoManager.java [hunk 13]
```diff
@@ -44,18 +48,30 @@
...
+  /** File path to the int to string enumeration file. */
+  private static final String INT_TO_STRING_ENUMERATION_FILE_NAME = "/usr/EnumerationMapping.json";
+
   /**
    * Populate the tag information list by using an Ewon Export Block Descriptor and parsing the
    * response.
    *
    * @throws IOException if EDB fails
    * @throws TagInfoBufferException if line from var_lst exceeds max capacity
+   * @throws JSONException if string enumeration JSON parse fails
```
Comment (alexjhawk) [inline src/com/hms_networks/americas/sc/taginfo/TagInfoManager.java:66]: You add the `@throws JSONException` in many places throughout this class, but the messages are not consistent. Most of them say `if int to string enumeration JSON parse fails` so this one should be updated to that, and any other inconsistent occurrences.

File: src/com/hms_networks/americas/sc/taginfo/package.html (hunk 3 above)
Comment (alexjhawk) [reply src/com/hms_networks/americas/sc/taginfo/package.html:11]: extra line was re-added/still needs to be removed

File: src/com/hms_networks/americas/sc/taginfo/TagInfoEnumeratedIntToString.java [hunk 14]
```diff
@@ -0,0 +1,80 @@
+package com.hms_networks.americas.sc.taginfo;
+
+import java.util.ArrayList;
+
+public class TagInfoEnumeratedIntToString extends TagInfo {
```
Comment (alexjhawk) [inline src/com/hms_networks/americas/sc/taginfo/TagInfoEnumeratedIntToString.java:5]: class missing javadoc. 

You can probably just copy/paste the javadocs from the regular TagInfo class and just tweak it for this.

File: src/com/hms_networks/americas/sc/taginfo/TagInfoEnumeratedIntToString.java [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+      String name,
+      boolean historicalLogEnabled,
+      boolean realTimeLogEnabled,
+      boolean isInGroupA,
+      boolean isInGroupB,
+      boolean isInGroupC,
+      boolean isInGroupD,
+      TagType type,
+      String[] enumeratedStringValueMapping) {
+
```
Comment (alexjhawk) [inline src/com/hms_networks/americas/sc/taginfo/TagInfoEnumeratedIntToString.java:35]: remove extra line before `super(`

**Review by alexjhawk — CHANGES_REQUESTED** [review]
A commit with version number updates and release commit with changelog updates should be added (if not this PR, to another one right after this)

**Comment by oliver-walker-hms** [comment]
@alexjhawk the next pr will have a release commit in it

File: src/com/hms_networks/americas/sc/taginfo/TagInfoEnumeratedIntToString.java (hunk 15 above)
Comment (alexjhawk) [reply src/com/hms_networks/americas/sc/taginfo/TagInfoEnumeratedIntToString.java:42]: Extra (and inconsistent) blank line still present: https://github.com/hms-networks/sc-flexy-tag-info-lib/blob/e47446ef7f3077f91debc313da9f45140fd2aa83/src/com/hms_networks/americas/sc/taginfo/TagInfoEnumeratedIntToString.java#L42

**Review by alexjhawk — CHANGES_REQUESTED** [review]

File: src/com/hms_networks/americas/sc/taginfo/TagInfoManager.java [hunk 16]
```diff
@@ -184,8 +200,12 @@ private static synchronized void rebuildInitialTagInfoListWithGaps() {
...
    *
    * @param line EBD line string
+   * @throws JSONException if int to string enumeration JSON parse fails
+   * @throws IOException if int to string enumeration file read fails
    */
-  private static synchronized void processTagListEBDLine(String line) {
+  private static synchronized void processTagListEBDLine(String line)
+      throws IOException, JSONException {
+
     /*
```
Comment (alexjhawk) [inline src/com/hms_networks/americas/sc/taginfo/TagInfoManager.java:209]:
```suggestion
  private static synchronized void processTagListEBDLine(String line)
      throws IOException, JSONException {
    /*
```

extra blank line added -- remove

**Review by alexjhawk — CHANGES_REQUESTED** [review]

**Review by alexjhawk — APPROVED** [review]

File: src/com/hms_networks/americas/sc/taginfo/TagInfoManager.java [hunk 17]
```diff
@@ -282,22 +301,194 @@ private static synchronized void processTagListEBDLine(String line) {
...
+       * for the JSON object index.
+       */
+      final int tagIntToStringEnumStringKeyIndex = 0;
+      for (int jsonArrayIndex = 0; jsonArrayIndex < length; jsonArrayIndex++) {
+        String currentKey =
+            (String)
+                ((JSONObject) tagIntToStringMappingsJsonArray.get(jsonArrayIndex))
+                    .names()
+                    .get(tagIntToStringEnumStringKeyIndex);
+        int current = Integer.parseInt(currentKey);
```
Comment (TomKimsey) [inline src/com/hms_networks/americas/sc/taginfo/TagInfoManager.java:480]: isnt there a NumberFormatException that can happen here?

File: src/com/hms_networks/americas/sc/taginfo/TagInfoManager.java [hunk 18]
```diff
@@ -282,22 +301,194 @@ private static synchronized void processTagListEBDLine(String line) {
...
+      final int length = tagIntToStringMappingsJsonArray.length();
+      int biggestIndex = 0;
+
+      /* Int to string enumerations are held in a JSON array of one key/value pair for each index in the array.
+       * To get the key of a key value pair, get the first(only) item returned by JSONObject.names()
+       * for the JSON object index.
+       */
+      final int tagIntToStringEnumStringKeyIndex = 0;
+      for (int jsonArrayIndex = 0; jsonArrayIndex < length; jsonArrayIndex++) {
+        String currentKey =
```
Comment (TomKimsey) [inline src/com/hms_networks/americas/sc/taginfo/TagInfoManager.java:475]: why store this if its not really going to be used, seems like this should just be 
```
int currentKey = Integer.parseInt(....)
```

If you want to keep it seperate just to keep the line length down there should be better naming scheme than currentKey and current (which are the same information, just different types.)

**Review by TomKimsey — CHANGES_REQUESTED** [review]

File: src/com/hms_networks/americas/sc/taginfo/TagInfoManager.java (hunk 18 above)
Comment (oliver-walker-hms) [reply src/com/hms_networks/americas/sc/taginfo/TagInfoManager.java]: changed the name and changed to not store the string

File: src/com/hms_networks/americas/sc/taginfo/TagInfoManager.java (hunk 17 above)
Comment (oliver-walker-hms) [reply src/com/hms_networks/americas/sc/taginfo/TagInfoManager.java]: true, I added this to the javadocs for both this and the calling function

File: src/com/hms_networks/americas/sc/taginfo/package.html [hunk 19]
```diff
@@ -3,8 +3,12 @@
 Ewon Flexy tag information library created by the HMS Networks MU Americas Solution Center.
 This library is dependent on the HMS Networks MU Americas Solution Center library for strings,
 found at https://www.github.com/hms-networks/sc-string-lib.
+This library is dependent on the HMS Networks MU Americas Solution Center library for file
+utilities, found at https://github.com/hms-networks/sc-flexy-file-util-lib.
+This library is dependent on the HMS Networks MU Americas Solution Center library for JSON
+management, found at https://github.com/hms-networks/sc-flexy-json-lib.
```
Comment (alexjhawk) [inline src/com/hms_networks/americas/sc/taginfo/package.html:9]: Also should be reflected in the README

**Review by alexjhawk — APPROVED** [review]

File: src/com/hms_networks/americas/sc/taginfo/package.html (hunk 19 above)
Comment (alexjhawk) [reply src/com/hms_networks/americas/sc/taginfo/package.html:9]: This can be added to the README as part of the pull request which will release this version.

File: src/com/hms_networks/americas/sc/taginfo/TagInfoEnumeratedIntToString.java [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+      TagType type,
+      String[] enumeratedStringValueMapping) {
+    super(id, name, historicalLogEnabled, realTimeLogEnabled, tagGroups, type);
+    this.enumeratedStringValueMapping = enumeratedStringValueMapping;
+  }
+
+  /**
+   * Get the enumerated int to String value mapping for this tag.
+   *
+   * @return the enumerated int to String value mapping for this tag.
```
Comment (it-hms) [inline src/com/hms_networks/americas/sc/taginfo/TagInfoEnumeratedIntToString.java:81]: inconsistent capitalization

**Review by it-hms — CHANGES_REQUESTED** [review]

File: src/com/hms_networks/americas/sc/taginfo/TagInfoEnumeratedIntToString.java (hunk 20 above)
Comment (alexjhawk) [reply src/com/hms_networks/americas/sc/taginfo/TagInfoEnumeratedIntToString.java]: fixed

File: src/com/hms_networks/americas/sc/taginfo/TagInfoEnumeratedIntToString.java [hunk 21]
```diff
@@ -0,0 +1,88 @@
+package com.hms_networks.americas.sc.taginfo;
+
+import java.util.ArrayList;
+
+/**
+ * Class to hold information about a tag and its configuration when the tag is an {@code int} to
+ * {@link String} enumeration tag.
```
Comment (TomKimsey) [inline src/com/hms_networks/americas/sc/taginfo/TagInfoEnumeratedIntToString.java:7]: Should its be "IntToStringEnumeration" instead of  "{@code int} to {@link String} enumeration"?

Comment (alexjhawk) [reply src/com/hms_networks/americas/sc/taginfo/TagInfoEnumeratedIntToString.java]: updated to better reflect that this aligns with TagType#INTEGER_MAPPED_STRING

**Review by TomKimsey — APPROVED** [review]

**Review by it-hms — APPROVED** [review]
